File: src/shesasyncer/evidence/ctc.py
from typing import Sequence
import math

from .acoustic import AcousticFrame


def ctc_viterbi_alignment(
    phonemes: Sequence[str],
    frames: Sequence[AcousticFrame],
    *,
    blank_token: str = "<blank>",
) -> tuple[dict, ...]:
    """Find the highest-scoring CTC path for trusted phonemes.

    The decoder uses the CTC blank explicitly, so silence can remain
    unassigned and repeated phonemes require the blank separation mandated by
    CTC. Scores are expected to be log-probabilities or another additive score.
    """
    tokens = [str(p).strip() for p in phonemes if str(p).strip()]
    if not tokens or not frames:
        return ()

    labels = [blank_token]
    for token in tokens:
        labels.extend((token, blank_token))

    n_states = len(labels)
    n_frames = len(frames)
    neg_inf = float("-inf")
    dp = [[neg_inf] * n_states for _ in range(n_frames)]
    back = [[-1] * n_states for _ in range(n_frames)]

    dp[0][0] = float(frames[0].scores.get(blank_token, neg_inf))
    if n_states > 1:
        dp[0][1] = float(frames[0].scores.get(tokens[0], neg_inf))

    for t in range(1, n_frames):
        scores = frames[t].scores
        for state, label in enumerate(labels):
            emit = float(scores.get(label, neg_inf))
            if emit == neg_inf:
                continue

            best_score = dp[t - 1][state]
            best_prev = state

            if state > 0 and dp[t - 1][state - 1] > best_score:
                best_score = dp[t - 1][state - 1]
                best_prev = state - 1

            if (
                state > 1
                and label != blank_token
                and label != labels[state - 2]
                and dp[t - 1][state - 2] > best_score
            ):
                best_score = dp[t - 1][state - 2]
                best_prev = state - 2

            if best_score != neg_inf:
                dp[t][state] = best_score + emit
                back[t][state] = best_prev

    final_candidates = [n_states - 1]
    if n_states > 1:
        final_candidates.append(n_states - 2)
    final_state = max(final_candidates, key=lambda state: dp[-1][state])
    if dp[-1][final_state] == neg_inf:
        return ()

    states = [final_state]
    for t in range(n_frames - 1, 0, -1):
        previous = back[t][states[-1]]
        if previous < 0:
            return ()
        states.append(previous)
    states.reverse()

    if len(frames) > 1:
        deltas = [max(0.0, frames[i + 1].time - frames[i].time) for i in range(n_frames - 1)]
        hop = sorted(deltas)[len(deltas) // 2] or 0.01
    else:
        hop = 0.01

    spans = []
    start = None
    current_state = None
    for i, state in enumerate(states + [-1]):
        label = labels[current_state] if current_state is not None else None
        if state != current_state:
            if current_state is not None and label != blank_token and start is not None:
                frame_scores = [
                    float(frames[k].scores.get(label, neg_inf))
                    for k in range(start, i)
                ]
                finite_scores = [score for score in frame_scores if score != neg_inf]
                if finite_scores:
                    mean_score = sum(finite_scores) / len(finite_scores)
                    confidence = math.exp(mean_score) if mean_score <= 0 else 1.0 - math.exp(-mean_score)
                    end_time = frames[i].time if i < n_frames else frames[-1].time + hop
                    spans.append({
                        "token_index": current_state // 2,
                        "phoneme": label,
                        "start": frames[start].time,
                        "end": end_time,
                        "confidence": max(0.0, min(1.0, confidence)),
                    })
            start = i if state >= 0 else None
            current_state = state if state >= 0 else None

    return tuple(spans)

File: src/shesasyncer/evidence/ctc.js
export function ctcViterbiAlignment(phonemes, frames, blankToken = '<blank>') {
  const tokens = phonemes.map(String).map(x => x.trim()).filter(Boolean); if (!tokens.length || !frames.length) return [];
  const labels = [blankToken]; for (const token of tokens) labels.push(token, blankToken);
  const states = labels.length, neg = -Infinity, dp = Array.from({ length: frames.length }, () => Array(states).fill(neg)), back = Array.from({ length: frames.length }, () => Array(states).fill(-1));
  dp[0][0] = Number(frames[0].scores[blankToken] ?? neg); if (states > 1) dp[0][1] = Number(frames[0].scores[tokens[0]] ?? neg);
  for (let t = 1; t < frames.length; t++) for (let s = 0; s < states; s++) {
    const emit = Number(frames[t].scores[labels[s]] ?? neg); if (emit === neg) continue;
    let best = dp[t - 1][s], prev = s;
    if (s > 0 && dp[t - 1][s - 1] > best) { best = dp[t - 1][s - 1]; prev = s - 1; }
    if (s > 1 && labels[s] !== blankToken && labels[s] !== labels[s - 2] && dp[t - 1][s - 2] > best) { best = dp[t - 1][s - 2]; prev = s - 2; }
    if (best !== neg) { dp[t][s] = best + emit; back[t][s] = prev; }
  }
  const finals = [states - 1, ...(states > 1 ? [states - 2] : [])], final = finals.reduce((a, b) => dp.at(-1)[b] > dp.at(-1)[a] ? b : a);
  if (dp.at(-1)[final] === neg) return [];
  const path = [final]; for (let t = frames.length - 1; t > 0; t--) { const p = back[t][path.at(-1)]; if (p < 0) return []; path.push(p); } path.reverse();
  const deltas = frames.slice(1).map((f, i) => Math.max(0, f.time - frames[i].time)).sort((a, b) => a - b), hop = deltas.length ? deltas[Math.floor(deltas.length / 2)] || 0.01 : 0.01;
  const spans = []; let start = null, current = null;
  for (let i = 0; i <= path.length; i++) { const state = i < path.length ? path[i] : -1, label = current == null ? null : labels[current];
    if (state !== current) {
      if (current != null && label !== blankToken && start != null) {
        const scores = frames.slice(start, i).map(f => Number(f.scores[label] ?? neg)).filter(Number.isFinite);
        if (scores.length) { const mean = scores.reduce((a, b) => a + b, 0) / scores.length; const conf = mean <= 0 ? Math.exp(mean) : 1 - Math.exp(-mean); spans.push({ tokenIndex: Math.floor(current / 2), phoneme: label, start: frames[start].time, end: i < frames.length ? frames[i].time : frames.at(-1).time + hop, confidence: Math.max(0, Math.min(1, conf)) }); }
      }
      start = state >= 0 ? i : null; current = state >= 0 ? state : null;
    }
  }
  return spans;
}
